cupy/functional/piecewise.py

```diff
@@ -0,0 +1,59 @@
+import cupy
+
+from cupy import core
+
+_piecewise_krnl = core.ElementwiseKernel(
+    'U condlist, S funclist',
+    'T y',
+    ' if(condlist) y = funclist',
+    'piecewise_kernel'
+)
+
+
+def piecewise(x, condlist, funclist):
+    """Evaluate a piecewise-defined function.
+
+        Args:
+            x (cupy.ndarray): input domain
+            condlist (cupy.ndarray or list):
+                conditions list ( boolean arrays or boolean scalars).
+                Each boolean array/ scalar corresponds to a function
+                in funclist. Length of funclist is equal to that
+                of condlist. If one extra function is given, it is used
+                as otherwise condition
+            funclist (cupy.ndarray or list): list of scalar functions.
+
+        Returns:
+            cupy.ndarray: the result of calling the functions in funclist
+                on portions of x defined by condlist.
+
+        .. warning::
+
+            This function currently doesn't support callable functions,
+            args and kw parameters.
+
+        .. seealso:: :func:`numpy.piecewise`
+        """
+
+    if any(callable(item) for item in funclist):
+        raise NotImplementedError(
+            'Callable functions are not supported currently')
+    if cupy.isscalar(condlist) or (not isinstance(
+            condlist[0], (list, cupy.ndarray)) and x.ndim != 0):
```

Comment:
I'm wondering if this condition is correct. Please add tests for non-list cupy.ndarray.

Reply:
Could you add a test for `not isinstance(condlist[0], (list, cupy.ndarray)) and x.ndim != 0)` condition?

Reply:
You use two kinds of conditions here: `cupy.isscalar(...)` and `isinstance(..., (list, cupy.ndarray)`. What is the difference between them?

Reply:
> Could you add a test for `not isinstance(condlist[0], (list, cupy.ndarray)) and x.ndim != 0)` condition?

``` test_scalar_condition ``` already tests that

Reply:
``` cupy.isscalar(condlist) ``` is used when for scalar condition as in ```test_scalar_condition``` 
```not isinstance(condlist[0], (list, cupy.ndarray)) and x.ndim != 0) ``` is used for multidimensional array input with condition list of size equal to shape[0] of input as in ``` test_piecewise_diff_shapes1, 2, 3, 4 ```

Reply:
I responded by sending the last fix but forgot to explain it here . Sorry for that :bow:

Reply:
Why did you write as `not isinstance(condlist[0], (list, cupy.ndarray))`? What is the difference from `cupy.isscalar(condlist[0])`?

Reply:
cupy.isscalar(condlist[0]) ≠ not isinstance(condlist[0], (list, cupy.ndarray)) because of tuple cases.
```
x = np.array([[0,1,4],[1,2,5],[3,4,3],[3,4,9]])
condlist = [(True, False, False), (False, True, True), (False, False, False), (False, True, True)]
>>> np.isscalar(condlist[0]))
False
>>> not isinstance(condlist[0], (list, np.ndarray)))
True
```
But I still need to handle it and add a test case